### PR TITLE
Kluent should only have an opinion on test runners

### LINKED ISSUE #4: Junit 5
Once junit 5 is released in a stable manner there should be a Kluent version targeting it.
The goal is to not break code using the current Kluent version, which can be accomplished if junit 5 is backwards compatible.

### PR DESCRIPTION
**Pull request by javatarz**
Removes dependency from `JUnit`. Project internally uses `kotlintest`

Users can choose their own test runners ([JUnit 4](http://junit.org/junit4/), [JUnit 5](http://junit.org/junit5/) or [Spek](http://spekframework.org/)). Kluent provides a fluent wrapper around [kotlintest](https://github.com/kotlintest/kotlintest).

**Note:**

- Major version has been bumped from 1.x to 2.x

### CODE REVIEW

**Comment by javatarz** [comment]
Added another commit ([d47b891](https://github.com/MarkusAmshove/Kluent/pull/72/commits/d47b89144e4012124c2151040c97be15c6a96052)) which removes JUnit 5 dependencies and only uses an assertion library.

Since Kluent is an assertion library, it feels weird that it brings in a test runner (JUnit) with itself. Moving Kluent 2.0 to use assertJ (another assertion library). Users can choose their test runners (JUnit 4, JUnit 5 or Spek) themselves.

**Comment by MarkusAmshove** [comment]
Wow, this is huge, thank you!

I'm not familiar with assertJ or opentest4j, how would I have to use Kluent in the future in my project? Is it enough to import Kluent and a runner of my choice and it "just works"?

I had the JUnit dependency because this was just a little side libary to depend upon for me ;-)

I would definitely take this as a major version bump to 2.0 and set all future additions ontop of it, therefore not maintain 1.x any further (aside from bug fixes maybe).

Maybe we can get some more feedback from @goreRatzete, @Egorand and @guenhter who also did a lot of contribution to this project.

This week is filled with work at my end, but I'll be able to have a closer look on sunday.

**Comment by javatarz** [comment]
# OpenTest4J
[OpenTest4j](https://github.com/ota4j-team/opentest4j) is a standard for how Java (or in this case JVM based projects throw test exceptions. AssertJ and a bunch of other frameworks already support it. OpenTest4J was created due to a [JUnit 5 initiative](https://github.com/junit-team/junit5/issues/12). You can read more about htis on the [OpenTest4j website](https://github.com/ota4j-team/opentest4j) .

In summary, using OpenTest4J changes nothing. You'd already have that as a dependency when you move to JUnit 5.

# Breaking changes in Kluent 2.x
You need to include your own test runner. You could choose JUnit 4, JUnit 5 or TestNG (to name a few examples.

JUnit 5 doesn't mind you using [third party assertion libraries](http://junit.org/junit5/docs/current/user-guide/#writing-tests-assertions-third-party). In fact, it sort-of recommends it for larger projects that need matchers for assertions.

File: README.md [hunk 1]
```diff
@@ -4,19 +4,17 @@
 
 It uses the [Infix-Notations](https://kotlinlang.org/docs/reference/functions.html#infix-notation "Infix-Notation") and [Extension Functions](https://kotlinlang.org/docs/reference/extensions.html#extension-functions "Extension Functions") of Kotlin to provide a fluent wrapper around the JUnit-Asserts and Mockito.
 
- [ ![Download](https://api.bintray.com/packages/markusamshove/maven/Kluent/images/download.svg) ](https://bintray.com/markusamshove/maven/Kluent/_latestVersion) [![Build Status](https://travis-ci.org/MarkusAmshove/Kluent.svg?branch=master)](https://travis-ci.org/MarkusAmshove/Kluent)
+Kluent 1.x uses JUnit 4 and Kluent 2.x uses JUnit 5. For more, read through the [Changelog and Documentation](https://markusamshove.github.io/Kluent/)
 
-
-# [Changelog](https://markusamshove.github.io/Kluent/)
-# [Documentation](https://markusamshove.github.io/Kluent/)
+[![Download](https://api.bintray.com/packages/markusamshove/maven/Kluent/images/download.svg)](https://bintray.com/markusamshove/maven/Kluent/_latestVersion) [![Build Status](https://travis-ci.org/MarkusAmshove/Kluent.svg?branch=master)](https://travis-ci.org/MarkusAmshove/Kluent)
```
Comment (Egorand) [inline README.md]: This should be in a separate PR

File: README.md [hunk 2]
```diff
@@ -4,19 +4,17 @@
...
 
 ----------
 
-# Include it via gradle/maven
+# Usage with dependency managers
```
Comment (Egorand) [inline README.md]: Would be great to have all edits/improvements in a separate PR

File: build.gradle [hunk 3]
```diff
@@ -36,11 +36,13 @@ repositories {
 }
 
 dependencies {
-    compile 'junit:junit:4.12'
+    compile 'org.assertj:assertj-core:3.8.0'
```
Comment (Egorand) [inline build.gradle]: As far as I recall, AssertJ 3.x targets Java 8, can it be used with Android? Although current Android toolchain generally supports Java 8, the best practice is to compile libraries against Java 6 if you're targeting Android.

File: build.gradle [hunk 4]
```diff
@@ -36,11 +36,13 @@ repositories {
 }
 
 dependencies {
-    compile 'junit:junit:4.12'
+    compile 'org.assertj:assertj-core:3.8.0'
+    compile 'org.opentest4j:opentest4j:1.0.0'
```
Comment (Egorand) [inline build.gradle]: I don't like dependencies on `assertj` and `opentest4j`, a more straightforward solution would be to depend on `kotlin-test` and use its assertions.

File: build.gradle (hunk 3 above)
Comment (guenhter) [reply build.gradle]: Yes, AssertJ targets Java8. Nicely spotted @Egorand 
I guess this is no real option to just drop the support for the whole Android World.

Comment (javatarz) [reply build.gradle]: Good call on the Android dependency. I was unaware of the Java 6 limitation.

We could move to [AssertJ 1.x for Java 6](https://joel-costigliola.github.io/assertj/assertj-core.html) though Core seems to suggest that you could use 2.x/3.x with Java6Assertions as the entry point. Honestly, we're not using any of the advanced features of AssertJ and moving to 1.x wouldn't be the worst thing.

File: build.gradle (hunk 4 above)
Comment (javatarz) [reply build.gradle]: I'll try moving to `kotlin-test`. I'm guessing (and hoping it works well on the android space) but I'm unsure.

If something goes wrong, we could move to `assertj-core:1.x` (or using `Java6Assertions` as the entry point which seemingly provides Android support)

File: build.gradle (hunk 3 above)
Comment (guenhter) [reply build.gradle]: You should just aim for a lib (in that case for a branch), which is still maintained. Or at least accepts pull requests for that branch. With assertJ, it not seems that the 1.x branch is development. I'm not sure if PR's to the 1.x would actually make it to the repo and to a new version of the 1.x stream.
That's quite sad, because I also really like AssertJ but IMO branch 2.x is no option (because of java 8) and 1.x seams pretty dead.

File: build.gradle (hunk 4 above)
Comment (MarkusAmshove) [reply build.gradle]: I like the idea of kotlin-test, since it is built in by kotlin. By deprecating Mocking and building upon kotlin-test we might be able to get platform independent and therefore support all platforms of kotlin (which JetBrains implements the library for).
Maybe that's a nice  goal for a 2.0 release. That way we would also get independent on junit (which this PR aims to do)

Comment (Egorand) [reply build.gradle]: Exactly, there's no reason why Kluent can't be used with Kotlin JS or Kotlin Native in the future.

**Comment by javatarz** [comment]
@MarkusAmshove @Egorand @guenhter: Moved to `kotlin-test`. I've held on to the `opentest4j` dependency.

[The project](https://github.com/ota4j-team/opentest4j/blob/master/build.gradle) is the new standard for assertion exceptions and should be followed by every JVM based project. It's also [Java 6](https://github.com/ota4j-team/opentest4j/blob/7f61828daa7dffd5c895f3fbbdaa189dee2378eb/build.gradle#L42) compliant so it works for Android.
It might not work for Kotlin-JS/Kotlin-Native but we'd need more test cases to prove that.

PS: We should add a test to ensure that the entire library is Java 6 compatible for android. It'd make it easier for contributors to provide changes in the future 😄

File: src/main/kotlin/org/amshove/kluent/Basic.kt [hunk 5]
```diff
@@ -1,31 +1,32 @@
 package org.amshove.kluent
 
-import org.junit.Assert.*
 import kotlin.reflect.KClass
+import kotlin.test.assertEquals
+import kotlin.test.assertNotEquals
 
-infix fun Any?.shouldEqual(theOther: Any?) = assertEquals(theOther, this)
+infix fun Any?.shouldEqual(theOther: Any?) = assertEquals(this, theOther)
```
Comment (Egorand) [inline src/main/kotlin/org/amshove/kluent/Basic.kt]: Parameter positions matter, first is `expected`, second is `actual`, so the correct invocation is `assertEquals(theOther, this)`

File: src/main/kotlin/org/amshove/kluent/Collections.kt [hunk 6]
```diff
@@ -1,26 +1,26 @@
...
+infix fun <T> Array<T>.shouldContainNone(things: Array<T>) = assert(this.none { things.contains(it) }, {"Expected $this to contain none of $things"})
 
 infix fun <T> Array<T>.shouldContainAll(things: Array<T>) = things.forEach { shouldContain(it) }
 
 infix fun <T> Array<T>.shouldNotContain(theThing: T) = if (!this.contains(theThing)) Unit else fail("$this should not contain $theThing", "the Array to not contain $theThing", join(this))
 
 infix fun <T> Array<T>.shouldNotContainAny(things: Array<T>) = things.forEach { shouldNotContain(it) }
 
-infix fun <T> Array<T>?.shouldEqual(theOther: Array<T>?) = assertArrayEquals(theOther, this)
+infix fun <T> Array<T>?.shouldEqual(theOther: Array<T>?) = assert(Arrays.equals(this, theOther))
```
Comment (Egorand) [inline src/main/kotlin/org/amshove/kluent/Collections.kt]: I think there should be a way to compare arrays in plain Kotlin, without the `java.util.Arrays` import

File: src/main/kotlin/org/amshove/kluent/Exceptions.kt [hunk 7]
```diff
@@ -9,9 +9,11 @@ infix fun <T : Throwable> (() -> Any?).shouldThrow(expectedException: KClass<T>)
         fail("There was an Exception expected to be thrown, but nothing was thrown", "$expectedException", "None")
     } catch (e: Throwable) {
         @Suppress("UNCHECKED_CAST")
-        if (e.isA(ComparisonFailure::class)) throw e
-        else if (e.isA(expectedException)) return ExceptionResult(e as T)
-        else throw ComparisonFailure("Expected ${expectedException.javaObjectType} to be thrown", "${expectedException.javaObjectType}", "${e.javaClass}")
+        return when {
+            e.isA(AssertionError::class) -> throw e
+            e.isA(expectedException) -> ExceptionResult(e as T)
+            else -> throw AssertionFailedError("Expected ${expectedException.javaObjectType} to be thrown", "${expectedException.javaObjectType}", "${e.javaClass}")
```
Comment (Egorand) [inline src/main/kotlin/org/amshove/kluent/Exceptions.kt]: Can this be replaced with `fail()` from `kotlin-test` to get rid of the opentest4j dependency?

File: build.gradle [hunk 8]
```diff
@@ -12,7 +12,7 @@ buildscript {
 }
 
 group 'org.amshove.kluent'
-version '1.30'
+version '2.0'
```
[the remark this answers is not in the document]
Comment (javatarz) [reply build.gradle]: I guess. I'm used to [CD](https://www.continuousdelivery.com/) where every commit "goes to prod". I had (incorrectly) assumed at the time that there was a pipeline that would publish a new artifact to Maven Central/BinTray.

As I understand it now @MarkusAmshove publishes artifacts manually. If he'd like, I can remove this change from this PR and let him version things.

File: src/main/kotlin/org/amshove/kluent/Collections.kt (hunk 6 above)
Comment (javatarz) [reply src/main/kotlin/org/amshove/kluent/Collections.kt]: `arr1 contentEquals arr2` does exist but it doesn't work since the arrays are optional.

File: src/main/kotlin/org/amshove/kluent/Exceptions.kt (hunk 7 above)
Comment (javatarz) [reply src/main/kotlin/org/amshove/kluent/Exceptions.kt]: Is there a reason for removing `opentest4j`? As [I said earlier](https://github.com/MarkusAmshove/Kluent/pull/72#issuecomment-340747784), it's Java 6 compliant and *the industry standard* for the exception to be thrown by assertion libraries.

`kotlin-test`'s `fail` uses `AssertionError` which does not express the errors as richly.

File: src/main/kotlin/org/amshove/kluent/Collections.kt (hunk 6 above)
Comment (Egorand) [reply src/main/kotlin/org/amshove/kluent/Collections.kt]: You can go with a helper function:

```kotlin
fun arraysEqual(a1: Array<Int>?, a2: Array<Int>?): Boolean {
  if (a1 == null || a2 == null) return false
  return a1 contentEquals a2
}
```

The behavior is consistent with `Arrays.equals()`

File: src/main/kotlin/org/amshove/kluent/Exceptions.kt (hunk 7 above)
Comment (Egorand) [reply src/main/kotlin/org/amshove/kluent/Exceptions.kt]: Yeah, agree with the reasoning. Indeed, what I was driving to is reducing the dependency on Java classes and libraries, but as you rightly mentioned, this has to be a separate discussion. I'd be fine with opentest4j for now.

File: docs/README.md [hunk 9]
```diff
@@ -40,6 +40,9 @@ Some examples:
     "hello" `should not equal` "world"
 
 # Changelog
+## 2.0 (WIP)
```
[the remark this answers is not in the document]
Comment (javatarz) [reply docs/README.md]: If @MarkusAmshove wants to version bump himself ([thread](https://github.com/MarkusAmshove/Kluent/pull/72/files/58c16d3d26ae35bf8bfaf51e3eb81aa165b0b94c#r148158447)), I'll remove this change.

[the remark this answers is not in the document]
Comment (javatarz) [reply docs/README.md]: If @MarkusAmshove wants to version bump himself (thread), I'll remove this change.

**Comment by javatarz** [comment]
@Egorand: I've used contentEquals where possible. The 1 instance it wasn't, I have kept `Array.equals`.

I think it's better to use a **standard Java implementation** than rolling out **our own non-standard** `arrayEquals` method. The argument is the same as the people implementing `toString`, `equals` and `hashCode` methods for simple classes in Java instead of using Kotlin's data classes.

If you agree, can I have an approval on the PR?

File: docs/README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (MarkusAmshove) [reply docs/README.md]: I'm okay with this being in there, I think we've settled that we want to use a new major version to do some cleanup in form of testrunners and mocking. I'll open an issue to discuss a roadmap for 2.0. Thank you for all the effort :)

**Comment by MarkusAmshove** [comment]
I've opened an issue (#75) to discuss the future of Kluent.
I want to base the 2.x on your pull requests, but also want to be sure to have a discussion written down to make sure we think of everything :-)

**Comment by javatarz** [comment]
The PR is over a year old and the merge conflicts will get harder to fix as time goes on.

With the roadmap now finalized, should I re-do this work on master for v2.0? If fixes are required to the latest 1.x version, we can do that off branches for @MarkusAmshove to publish artifacts.

If this makes sense (looking for @MarkusAmshove to give the thumbs up on the version bump), I can go ahead and strart fixing this PR :)

**Comment by MarkusAmshove** [comment]
At the moment this is still blocked, because the different runner implementations depending on the platform `kotlin.test` runs on doesn't make it possible to handle `Exception`s at the moment (as we have to differentiate between exceptions from the testrunner).

Since there are a lot of conflicts as the repo structure has changed dramatically, I'll close this.

I have however annotated a lot of the build stuff with TODOs to know what to change when Kotlin multiplatform gets there :)
